Native application execution failure with @Retry on the class and private methods

### Describe the bug

Hi,

a native application fails to start with the following logs:
```
2025-04-23 17:43:09,812 ERROR [io.qua.run.Application] (main) Failed to start application: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:119)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:80)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:51)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:144)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
Caused by: jakarta.enterprise.inject.spi.DeploymentException: java.lang.NoSuchMethodException: TestRestryService.myPrivateMethod()
        at io.quarkus.smallrye.faulttolerance.runtime.SmallRyeFaultToleranceRecorder.createFaultToleranceOperation(SmallRyeFaultToleranceRecorder.java:44)
        at io.quarkus.runner.recorded.SmallRyeFaultToleranceProcessor$processFaultToleranceAnnotations1222864980.deploy_0(Unknown Source)
        at io.quarkus.runner.recorded.SmallRyeFaultToleranceProcessor$processFaultToleranceAnnotations1222864980.deploy(Unknown Source)
        ... 7 more
Caused by: java.lang.NoSuchMethodException: TestRestryService.myPrivateMethod()
        at java.base@21.0.6/java.lang.Class.checkMethod(DynamicHub.java:1078)
        at java.base@21.0.6/java.lang.Class.getDeclaredMethod(DynamicHub.java:1168)
        at io.smallrye.faulttolerance.autoconfig.MethodDescriptor.reflect(MethodDescriptor.java:28)
        at io.quarkus.smallrye.faulttolerance.runtime.SmallRyeFaultToleranceRecorder.createFaultToleranceOperation(SmallRyeFaultToleranceRecorder.java:31)
        ... 9 more
```

This is caused by the use of @Retry at class level in a class that contains a private method.

Thanks for the help.

### Expected behavior

The native application start properly.

### Actual behavior

The native application fails.

### How to Reproduce?

[code-with-quarkus-native-retry.zip](https://github.com/user-attachments/files/19871880/code-with-quarkus-native-retry.zip)

- build a native executable (with Docker `mvn install -Dnative -DskipTests -Dquarkus.native.container-build=true`)
- run the native app: ./target/code-with-quarkus-1.0.0-SNAPSHOT-runner

### Output of `uname -a` or `ver`

Ubuntu 24.04.2 LTS

### Output of `java -version`

openjdk version "21.0.6" 2025-01-21

### Mandrel or GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.21.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.7

### Additional information

_No response_

--- a/extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/FaultToleranceScanner.java
+++ b/extensions/smallrye-fault-tolerance/deployment/src/main/java/io/quarkus/smallrye/faulttolerance/deployment/FaultToleranceScanner.java
@@ -1,6 +1,7 @@
 package io.quarkus.smallrye.faulttolerance.deployment;
 
 import java.lang.annotation.Annotation;
+import java.lang.reflect.Modifier;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashSet;
@@ -100,6 +101,10 @@ final class FaultToleranceScanner {
             }
             if (method.isSynthetic()) {
                 // synthetic methods can't be intercepted
+                continue;
+            }
+            if (Modifier.isPrivate(method.flags())) {
+                // private methods can't be intercepted
                 continue;
             }
             if (annotationStore.hasAnnotation(method, io.quarkus.arc.processor.DotNames.NO_CLASS_INTERCEPTORS)

--- a/integration-tests/main/src/main/java/io/quarkus/it/faulttolerance/FaultToleranceTestResource.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/faulttolerance/FaultToleranceTestResource.java
@@ -12,6 +12,9 @@ public class FaultToleranceTestResource {
     @Inject
     Service service;
 
+    @Inject
+    SecondService secondService;
+
     @GET
     public String getName() {
         AtomicInteger counter = new AtomicInteger();
@@ -33,5 +36,11 @@ public class FaultToleranceTestResource {
         AtomicInteger counter = new AtomicInteger();
         String name = service.fallbackMethod(counter);
         return counter + ":" + name;
+    }
+
+    @GET
+    @Path("/hello")
+    public String hello() {
+        return secondService.publicHello();
     }
 }

--- a/integration-tests/main/src/main/java/io/quarkus/it/faulttolerance/SecondService.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/faulttolerance/SecondService.java
@@ -1,0 +1,17 @@
+package io.quarkus.it.faulttolerance;
+
+import jakarta.enterprise.context.ApplicationScoped;
+
+import org.eclipse.microprofile.faulttolerance.Retry;
+
+@ApplicationScoped
+@Retry
+public class SecondService {
+    public String publicHello() {
+        return "hello";
+    }
+
+    private String privateHello() {
+        return "hello";
+    }
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/FaultToleranceTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/FaultToleranceTestCase.java
@@ -35,5 +35,10 @@ public class FaultToleranceTestCase {
                 .given().baseUri(uri.toString() + "/fallback")
                 .when().get()
                 .then().body(is("1:fallback"));
+
+        RestAssured
+                .given().baseUri(uri.toString() + "/hello")
+                .when().get()
+                .then().body(is("hello"));
     }
 }

